src/Agent.Worker/TaskDefinitionManager.cs

```diff
@@ -0,0 +1,291 @@
+                Directory.Move(tempDirectory, destDirectory);
+                Trace.Info("Finished getting task.");
+            }
+            finally
+            {
+                try
+                {
+                    //sometimes the temp folder is not deleted -> wipe it
+                    if (Directory.Exists(tempDirectory))
+                    {
+                        Trace.Verbose("Deleting task temp folder: {0}", tempDirectory);
+                        Directory.Delete(tempDirectory, true);
+                    }
+                }
+                catch (Exception ex)
+                {
+                    //it is not critical if we fail to delete the temp folder
+                    Trace.Warning("Failed to delete temp folder '{0}'. Exception: {1}", tempDirectory, ex);
+                    executionContext.Warning(StringUtil.Loc("FailedDeletingTempDirectory0Message1", tempDirectory, ex.Message));
+                }
+            }
+        }
+
+        private string GetDirectory(TaskReference task)
+        {
+            ArgUtil.NotEmpty(task.Id, nameof(task.Id));
+            ArgUtil.NotNull(task.Name, nameof(task.Name));
+            ArgUtil.NotNullOrEmpty(task.Version, nameof(task.Version));
+            return Path.Combine(
+                IOUtil.GetTasksPath(HostContext),
+                $"{task.Name}_{task.Id}",
+                task.Version);
+        }
+
+        public async Task DownloadAsync(IExecutionContext executionContext, IEnumerable<TaskInstance> tasks)
+        {
+            ArgUtil.NotNull(executionContext, nameof(executionContext));
+            ArgUtil.NotNull(tasks, nameof(tasks));
+
+            //remove duplicate and disabled tasks
+            IEnumerable<TaskInstance> uniqueTasks =
+                from task in tasks
+                where task.Enabled
+                group task by new
+                    {
+                        task.Id,
+                        task.Version
+                    }
+                into taskGrouping
+                select taskGrouping.First();
+            foreach (TaskInstance task in uniqueTasks)
+            {
+                await DownloadAsync(executionContext, task);
+            }
+        }
+
+        public Definition Load(TaskReference task)
+        {
+            // Validate args.
+            Trace.Entering();
+            ArgUtil.NotNull(task, nameof(task));
+
+            // Initialize the definition wrapper object.
+            var definition = new Definition() { Directory = GetDirectory(task) };
+
+            // Deserialize the JSON.
+            string file = Path.Combine(definition.Directory, Constants.Path.TaskJsonFile);
+            Trace.Info($"Loading task definition '{file}'.");
+            string json = File.ReadAllText(file);
+            definition.Data = JsonConvert.DeserializeObject<DefinitionData>(json);
+
+            // Replace the macros within the handler data sections.
```

Comment:
The macro replacement for `$(currentDirectory)` is kind of ridiculous. But at least it's contained within the code in this file.